united-fare-finder.py

```diff
@@ -36,45 +36,96 @@
-elem.click()
-
-elem = driver.find_element(
-  By.CSS_SELECTOR,
-  "button.app-components-BookFlightForm-bookFlightForm__findFlightBtn--1lbFe")
-elem.submit()
-
-time.sleep(10)
-
-# scrape fare data
-page_source = driver.page_source
-flight_elements = parsePageSource(page_source)
-
-# write to file
-for i in range(len(flight_elements)):
-    for j in range(len(flight_elements[i])):
-        print(*flight_elements[i][j])
-    print()
+origin = "ORD"
+dest = [
+  "hnl", "slc", "dfw", "lax", "sea", "atl",
+  "mia", "mco", "sfo", "jfk", "ewr", "lga", "dca", "iad"
+]
+depDate = "03/01/2023"
+retDate = "03/10/2023"
+
+count = 0
+for dest in dest:
+
+    # input flight info in main page
+    if (count == 0):
+        elem = driver.find_element(By.ID, "bookFlightOriginInput")
+        elem.clear()
+        elem.send_keys(origin)
+
+        elem = driver.find_element(By.ID, "bookFlightDestinationInput")
+        elem.clear()
+        elem.send_keys(dest)
+
+        elem = driver.find_element(By.ID, "DepartDate")
+        elem.clear()
+        elem.send_keys(depDate)
+
+        elem = driver.find_element(By.ID, "ReturnDate")
+        elem.clear()
+        elem.send_keys(retDate)
+
+        elem = driver.find_element(
+          By.XPATH, '//*[@id="bookFlightForm"]/div[1]/div[1]/label')
+        elem.click()
+
+        elem = driver.find_element(
+          By.CSS_SELECTOR,
+          "button."
+          "app-components-BookFlightForm-bookFlightForm__findFlightBtn--1lbFe")
+        elem.submit()
+
+    # input flight info in subsequent pages
+    else:
+        elem = driver.find_element(
+          By.ID, "airport-autocomplete-.swap0.fromTrip")
+        elem.send_keys(Keys.CONTROL + "a")
+        elem.send_keys(origin)
+
+        elem = driver.find_element(
+          By.XPATH,
+          '//li[@id="airport-autocomplete-.swap0.fromTrip-menu-item-0"]//div')
+        elem.click()
+
+        elem = driver.find_element(By.ID, "airport-autocomplete-.swap0.toTrip")
+        elem.send_keys(Keys.CONTROL + "a")
+        elem.send_keys(dest)
+
+        elem = driver.find_element(
+          By.XPATH,
+          '//li[@id="airport-autocomplete-.swap0.toTrip-menu-item-0"]//div')
+        elem.click()
+
+        elemList = driver.find_elements(
+          By.XPATH, '//div[@class="DateInput DateInput_1"]/input')
+
+        elem = elemList[0]
+        elem.send_keys(Keys.CONTROL + "a")
+        elem.send_keys(depDate)
+
+        elem = elemList[1]
+        elem.send_keys(Keys.CONTROL + "a")
+        elem.send_keys(retDate)
+
+        elem = driver.find_element(
+          By.CSS_SELECTOR,
+          ".atm-c-btn.atm-c-btn--secondary")
+        elem.click()
+
+    time.sleep(15)
+
+    # scrape fare data
+    page_source = driver.page_source
+    soup = BeautifulSoup(page_source, "html.parser")
+    flight_elements = parsePageSource(soup)
+
+    # write to file
+    print(dest + " ********************************************************\n")
```

Comment:
the second iteration errored out for me:
```no such element: Unable to locate element: {"method":"xpath","selector":"//li[@id="airport-autocomplete-.swap0.fromTrip-menu-item-0"]//div"}```

Generally the more elements you have in a script the more brittle it is going to be. any code update that they do to the page has the potential to break it (esp when using XPath or CSS selector, Id is the safest)

I think the least brittle way to do this is to just iterate through the same script that we had working for 1, and do it for every destination.  Completely starting over on the homepage with a reload, that way we can be sure of the beginning state of the page.

Maybe we don't login on every iteration though, because login itself is probably the most error prone part of this process.

Reply:
noted!  I was trying to get a little fancy.  New PR request will have iteration through the original code, with a couple of tweaks.